google_guest_agent/instance_setup.go

```diff
@@ -190,6 +214,340 @@ func agentInit(ctx context.Context) {
+//processor 0. For this virtionet configuration, distributing IRQs to all
+//processors results in comparatively high cpu utilization and comparatively
+//low network bandwidth.
+//
+//For a multi-queue / MSI-X virtionet device, sets the IRQ affinities to the
+//per-IRQ affinity hint. The virtionet driver maps each virtionet TX (RX) queue
+//MSI-X interrupt to a unique single CPU if the number of TX (RX) queues equals
+//the number of online CPUs. The mapping of network MSI-X interrupt vector to
+//CPUs is stored in the virtionet MSI-X interrupt vector affinity hint. This
+//configuration allows network traffic to be spread across the CPUs, giving
+//each CPU a dedicated TX and RX network queue, while ensuring that all packets
+//from a single flow are delivered to the same CPU.
+//
+//For a gvnic device, set the IRQ affinities to the per-IRQ affinity hint.
+//The google virtual ethernet driver maps each queue MSI-X interrupt to a
+//unique single CPU, which is stored in the affinity_hint for each MSI-X
+//vector. In older versions of the kernel, irqblanace is expected to copy the
+//affinity_hint to smp_affinity; however, GCE instances disable irqbalance by
+//default. This script copies over the affinity_hint to smp_affinity on boot to
+//replicate the behavior of irqbalance.
+func setMultiQueue(totalCPUs int) error {
+	devices, err := filepath.Glob(virtioNetDevs)
+	if err != nil {
+		return err
+	}
+	for _, dev := range devices {
+		if err := enableMultiQueue(dev); err != nil {
+			logger.Warningf("Could not enable multi queue for %s.", dev)
+		}
+		if err = setQueueNumForDevice(dev); err != nil {
+			logger.Warningf("Could not set queue num for %s.", dev)
+		}
+	}
+	// Set smp_affinity properly for gvnic queues. '-ntfy-block.' is unique to gve and will not affect virtio queues.
+	if err = setSMPAffinityForGVNIC(); err != nil {
+		logger.Warningf("Could not set smp_affinity for gvnic.")
+	}
+
+	if err = configureTransmitPacketSteering(totalCPUs); err != nil {
+		logger.Warningf("Could not set transmit packet steering.")
+	}
+	return nil
+}
+
+func setSMPAffinityForGVNIC() error {
+	irqDirs, err := filepath.Glob(irqDirPath)
+	if err != nil {
+		return err
+	}
+	for _, irq := range irqDirs {
+		blocks, err := filepath.Glob(irq + "/*-ntfy-block.*")
+		if err != nil {
+			return err
+		}
+		if len(blocks) > 0 && isFile(irq+"/affinity_hint") {
+			if err = copyFile(irq+"/affinity_hint", irq+"/smp_affinity"); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func configureTransmitPacketSteering(totalCPUs int) error {
+	XPS, err := filepath.Glob(xpsCPU)
+	if err != nil {
+		return err
+	}
+
+	// If we have more CPUs than queues, then stripe CPUs across tx affinity as CPUNumber % queue_count.
+	for _, q := range XPS {
+		numQueues := len(XPS)
+		r, _ := regexp.Compile(queueRegex)
+		var queueNum int
+		if r.MatchString(q) {
+			queueNum, err = strconv.Atoi(r.FindAllStringSubmatch(q, -1)[0][1])
```

Comment:
use FindStringSubmatch instead - all is if the entire expression may match multiple times

Reply:
Done